### PR TITLE
optimize: unlock branch synchronously.

### LINKED ISSUE #4580: rollback transaction should release the global lock before asynchronously deleting the branch transaction
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
如果大量分支事务被异步删除情况下,branch被异步删除,解锁等于也异步了,回滚情况下,异步删除线程是跟提交都是在一个线程池中,一个队列中,当一个回滚的事务在队列中被排得相对靠后,那么解锁等于慢了,势必影响吞吐,特别是热点数据本身可以在业务线程中被解锁,却被放到了队列中,我个人觉得回滚情况下不应该异步删除branch,起码也是得先解锁再异步,提交倒是无所谓,因为锁事前就被全部解开了

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
Just paste your stack trace here!
```

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅵ. Environment:

- JDK version :
- Seata version:  
- OS :
- Others:

### PR DESCRIPTION
**Pull request by pengten**
- [x] I have registered the PR [changes](https://github.com/seata/seata/tree/develop/changes).

### Ⅰ. Describe what this PR did
If `server.session.enable-branch-async-remove` is true, delete the branch asynchronously and unlock it synchronously.

### Ⅱ. Does this pull request fix one issue?

fixes  #4580

### CODE REVIEW

File: server/src/main/java/io/seata/server/session/SessionHelper.java [hunk 1]
```diff
@@ -294,6 +294,7 @@ public static Boolean forEach(Collection<BranchSession> sessions, BranchSessionH
      */
     public static void removeBranch(GlobalSession globalSession, BranchSession branchSession, boolean isAsync)
             throws TransactionException {
+        globalSession.unlockBranch(branchSession);
         if (Objects.equals(Boolean.TRUE, ENABLE_BRANCH_ASYNC_REMOVE) && isAsync) {
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/session/SessionHelper.java:298]: DELAY_HANDLE_SESSION  增加这个判断,不需要delay的不需要异步

File: server/src/main/java/io/seata/server/session/GlobalSession.java [hunk 2]
```diff
@@ -312,14 +312,18 @@ public void loadBranchs() {
...
         // because it's already unlocked in 'DefaultCore.commit()'
         if (status != Committing && status != CommitRetrying && status != AsyncCommitting) {
             if (!branchSession.unlock()) {
                 throw new TransactionException("Unlock branch lock failed, xid = " + this.xid + ", branchId = " + branchSession.getBranchId());
             }
         }
+    }
+
+    @Override
+    public void removeBranch(BranchSession branchSession) throws TransactionException {
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/session/GlobalSession.java:326]: remove branch要包含unlockbranch的动作,增加一个removeBranchAndUnlock的方法吧

File: server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java [hunk 3]
```diff
@@ -194,6 +190,16 @@ private DefaultCoordinator(RemotingServer remotingServer) {
...
+        // create branchRemoveExecutor
+        if (StoreConfig.getSessionMode() != StoreConfig.SessionMode.FILE) {
+            branchRemoveExecutor = new ThreadPoolExecutor(BRANCH_ASYNC_POOL_SIZE, BRANCH_ASYNC_POOL_SIZE,
+                    Integer.MAX_VALUE, TimeUnit.MILLISECONDS,
+                    new ArrayBlockingQueue<>(
+                            CONFIG.getInt(ConfigurationKeys.SESSION_BRANCH_ASYNC_QUEUE_SIZE, DEFAULT_BRANCH_ASYNC_QUEUE_SIZE)
+                    ), new NamedThreadFactory("branchSessionRemove", BRANCH_ASYNC_POOL_SIZE),
+                    new ThreadPoolExecutor.CallerRunsPolicy());
+        }
     }
```
Comment (pengten) [inline server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java:203]: @a364176773 watch this

**Review by a364176773 — APPROVED** [review]
LGTM @slievrly PTAL

File: server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java [hunk 4]
```diff
@@ -194,6 +190,18 @@ private DefaultCoordinator(RemotingServer remotingServer) {
         }
         this.remotingServer = remotingServer;
         this.core = new DefaultCore(remotingServer);
+
+        // create branchRemoveExecutor
+        if (StoreConfig.getSessionMode() != StoreConfig.SessionMode.FILE) {
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/coordinator/DefaultCoordinator.java:195]: 还要判断是否开启了异步化删除分支事务，否者也没必要创建线程池

**Review by xingfudeshi — APPROVED** [review]
LGTM

**Pull request merged**